idp-binding-service/src/main/java/io/mosip/idp/binding/services/KeyBindingServiceImpl.java

```diff
@@ -0,0 +1,320 @@
+				LocalDateTime.now(ZoneOffset.UTC));
+		if (!publicKeyRegistry.isPresent())
+			throw new IdPException(ErrorConstants.KEY_BINDING_NOT_FOUND);
+
+		List<String> supportedAuthFactors = null;
+		try {
+			supportedAuthFactors = objectMapper.readValue(publicKeyRegistry.get().getAuthFactors(), new TypeReference<List<String>>() {});
+		} catch (IllegalArgumentException | JsonProcessingException e) {
+			log.error("Failed to parse supported auth-factors", e);
+		}
+
+		//check if provided challenge auth-factor is the bound auth-factor-type for the provided individualId
+		if(CollectionUtils.isEmpty(supportedAuthFactors) || !supportedAuthFactors.contains(validateBindingRequest.getChallenge().getAuthFactorType()))
+			throw new IdPException(ErrorConstants.UNBOUND_AUTH_FACTOR);
+
+		X509Certificate x509Certificate = null;
+		try {
+			x509Certificate = (X509Certificate) keymanagerUtil.convertToCertificate(publicKeyRegistry.get().getCertificate());
+			//TODO should we validate the certificate?
+		} catch (KeymanagerServiceException e) {
+			log.error("Certificate parsing failed", e);
+			throw new IdPException(ErrorConstants.INVALID_CERTIFICATE);
+		}
+
+		switch (validateBindingRequest.getChallenge().getAuthFactorType()) {
+			case "WLA" :
+				verifyWLAToken(validateBindingRequest.getIndividualId(), validateBindingRequest.getChallenge().getChallenge(),
+						validateBindingRequest.getChallenge().getFormat(), publicKeyRegistry.get().getPublicKey(),
+						X509Util.x5t(x509Certificate));
+				break;
+			default:
+				throw new IdPException(UNKNOWN_BINDING_CHALLENGE);
+		}
+
+		return new ValidateBindingResponse(validateBindingRequest.getIndividualId(), validateBindingRequest.getTransactionId());
+	}
+
+	private void verifyWLAToken(String individualId, String wlaToken, String format, String publicKey, String thumbprint) {
+		if(format == null) { format = "JWT"; }
+
+		switch (format) {
+			case "JWT" :
+				try {
+					JWSKeySelector keySelector = new JWSVerificationKeySelector(JWSAlgorithm.RS256,
+							new ImmutableJWKSet(new JWKSet(RSAKey.parse(publicKey))));
+					JWTClaimsSetVerifier claimsSetVerifier = new DefaultJWTClaimsVerifier(new JWTClaimsSet.Builder()
+							.audience(validateBindingUrl)
+							.subject(individualId)
+							.build(), REQUIRED_WLA_CLAIMS);
+
+					ConfigurableJWTProcessor jwtProcessor = new DefaultJWTProcessor();
+					jwtProcessor.setJWSKeySelector(keySelector);
+					jwtProcessor.setJWTClaimsSetVerifier(claimsSetVerifier);
+					jwtProcessor.process(wlaToken, null); //If invalid throws exception
+
+					SignedJWT signedJWT = SignedJWT.parse(wlaToken);
+					if(signedJWT.getHeader().getX509CertThumbprint() == null ||
+							!signedJWT.getHeader().getX509CertThumbprint().toString().equals(thumbprint)) {
+						log.error("SHA-1 thumbprint is invalid {}", signedJWT.getHeader().getX509CertThumbprint());
+						throw new IdPException(ErrorConstants.INVALID_WLA_TOKEN);
+					}
```

Comment:
does is make sense to do this before the JWT validation or can we include this in the above verfiication itself, that way we don't parse the wla token 2 times

Reply:
Move the WLA validation to different helper class